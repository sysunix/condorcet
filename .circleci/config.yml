version: 2.1

commands:
  restore_cypress:
    steps:
      - restore_cache:
          key: v1-cypress-binary-{{ checksum "yarn.lock" }}

  restore_functions:
    steps:
      - restore_cache:
          key: v4-functions-dependencies-{{ checksum "functions/yarn.lock" }}

  checkout_and_restore_cache:
    steps:
      - checkout
      - restore_cache:
          key: v3-dependencies-{{ checksum "yarn.lock" }}

executors:
  node:
    docker:
      - image: cypress/base:12.14.0
    working_directory: ~/app

jobs:
  install:
    executor: node
    steps:
      - checkout_and_restore_cache
      - restore_functions
      - restore_cypress
      - run: yarn install
      - run: yarn cypress install
      - run: cd ~/app/functions && yarn install
      - save_cache:
          key: v3-dependencies-{{ checksum "yarn.lock" }}
          paths:
            - ./node_modules
      - save_cache:
          key: v4-functions-dependencies-{{ checksum "functions/yarn.lock" }}
          paths:
            - ~/app/functions/node_modules
      - save_cache:
          key: v1-cypress-binary-{{ checksum "yarn.lock" }}
          paths:
            - /root/.cache/Cypress

  build:
    executor: node
    steps:
      - checkout_and_restore_cache
      - run: yarn build
      - persist_to_workspace:
          root: ~/app
          paths:
            - ./dist

  lint:
    executor: node
    steps:
      - checkout_and_restore_cache
      - run: yarn lint

  test_unit:
    executor: node
    steps:
      - checkout_and_restore_cache
      - restore_functions
      - run: yarn test:unit --coverage
      - run: yarn codecov

  test_e2e:
    executor: node
    steps:
      - checkout_and_restore_cache
      - restore_cypress
      - attach_workspace:
          at: ~/app
      - run:
          command: yarn firebase serve --only hosting -p 8080
          background: true
      - run: echo $SERVICE_ACCOUNT > ./serviceAccount.json
      - run: yarn test:e2e:ci

  deploy:
    executor: node
    steps:
      - checkout_and_restore_cache
      - attach_workspace:
          at: ~/app
      - run: yarn firebase deploy --token "${FIREBASE_TOKEN}" --only hosting

  check_pr:
    executor: node
    steps:
      - checkout
      - run: cd tools/assistant && yarn install
      - run: PR_NUMBER=${CIRCLE_PR_NUMBER} node ./tools/assistant/scripts/callChecker.js

workflows:
  version: 2
  integration:
    jobs:
      - check_pr:
          filters:
            branches:
              only: /^(pull\/).*$/
      - install
      - build:
          requires:
            - install
      - lint:
          requires:
            - build
      - test_unit:
          requires:
            - build
      - deploy:
          requires:
            - lint
            - test_unit
            # - test_e2e
          filters:
            branches:
              only: master
