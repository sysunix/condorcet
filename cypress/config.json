{
    "TEST_UID": "Kb5XM9p2CPhGtiMqgdGVUO8H49R2",
    "FIREBASE_PROJECT_ID": "condorcet-4167c",
    "FIREBASE_API_KEY": "AIzaSyBl7KGWaRY0B-BQ7WhUwSnmFHj3f3OYF0U"
}
