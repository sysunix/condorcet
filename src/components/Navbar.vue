<template>
  <nav
    class="Menu fixed bottom-0 left-0 right-0 lg:top-0 lg:bottom-auto flex items-center justify-between flex-wrap bg-teal-500 p-6 z-50"
    :class="isMenuOpen ? 'Menu--show' : 'Menu--hide'"
  >
    <div class="flex items-center flex-shrink-0 text-white mr-6">
      <span class="font-semibold text-xl tracking-tight">Condorcet</span>
    </div>
    <div class="block lg:hidden">
      <button
        data-testid="burger-button"
        class="flex items-center px-3 py-2 border rounded text-teal-200 border-teal-400 hover:text-white hover:border-white"
        @click="toggleMenu"
      >
        <svg
          class="fill-current h-3 w-3"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>Menu</title>
          <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z" />
        </svg>
      </button>
    </div>
    <div class="justify-end w-full block lg:flex lg:items-center lg:w-auto">
      <div class="text-sm lg:flex-grow">
        <router-link
          data-testid="navlink-user-polls"
          to="/polls"
          class="block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white"
          @click.native="toggleMenu"
          >{{ $t("navigation.userPolls") }}</router-link
        >
        <router-link
          v-if="featureFlipping.publicPolls"
          data-testid="navlink-public-polls"
          to="/polls/public"
          class="self-end block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white"
          @click.native="toggleMenu"
          >{{ $t("navigation.publicPolls") }}</router-link
        >
        <router-link
          data-testid="navlink-poll-creation"
          to="/polls/new"
          class="self-end block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white"
          @click.native="toggleMenu"
          >{{ $t("navigation.createPoll") }}</router-link
        >
      </div>
      <div>
        <a
          href="#"
          class="inline-block text-sm px-4 py-2 leading-none border rounded text-white border-white hover:border-transparent hover:text-teal-500 hover:bg-white mt-4 lg:mt-0"
          @click="signOut"
          >{{ $t("navigation.logOut") }}</a
        >
      </div>
    </div>
  </nav>
</template>

<script>
import { mapState, mapActions } from "vuex";
import { signOut } from "../utils/authentication.js";
export default {
  computed: {
    ...mapState("app", ["notifications", "featureFlipping", "isMenuOpen"])
  },
  methods: {
    ...mapActions("user", ["clearUser"]),
    ...mapActions("app", ["addNotification", "toggleMenu"]),
    async signOut() {
      try {
        await signOut();
        this.clearUser();
        this.$router.push({ name: "authentication" });
      } catch (error) {
        this.addNotification({
          status: "error",
          message: "Il y a eu un probl√®me"
        });
      }
    }
  }
};
</script>

<style>
.Menu {
  transition: transform 300ms;
}

@media (max-width: 1024px) {
  .Menu--hide {
    transform: translateY(165px);
  }

  .Menu--show {
    transform: translateY(0);
  }
}
</style>
