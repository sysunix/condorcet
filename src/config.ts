export default {
  notificationTimeout: 3000,
  featureFlipping: {
    publicPolls: false,
    graphOfTheDuels: false,
    matrixOfTheDuels: false,
    condorcetRanking: true,
    uninominalRanking: true
  }
};
