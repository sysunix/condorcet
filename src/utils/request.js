import firebase, { db } from "../firebase";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db.collection("polls").add(pollWithTimestap);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(voteWithTimestap);
};
