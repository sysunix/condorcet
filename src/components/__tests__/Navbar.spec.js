import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Navbar from "../Navbar.vue";
import { signOut } from "../../utils/authentication";
import config from "../../config";
import { getTranslation } from "../../utils/test";

jest.mock("../../utils/authentication");

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Navbar.vue", () => {
  let wrapper;
  let appActions;
  let userActions;

  beforeEach(() => {
    jest.clearAllMocks();
    appActions = { addNotification: jest.fn() };
    userActions = { clearUser: jest.fn() };

    const store = new Vuex.Store({
      modules: {
        app: {
          namespaced: true,
          actions: appActions,
          state: {
            featureFlipping: config.featureFlipping
          }
        },
        user: { namespaced: true, actions: userActions }
      }
    });

    wrapper = shallowMount(Navbar, {
      localVue,
      store,
      stubs: ["router-link"],
      mocks: {
        $t: getTranslation
      }
    });
  });

  it("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should signout user", async () => {
    const signOutLink = wrapper.find("a");

    signOutLink.trigger("click");

    await expect(signOut).toBeCalled();
    expect(userActions.clearUser).toBeCalled();
  });
});
