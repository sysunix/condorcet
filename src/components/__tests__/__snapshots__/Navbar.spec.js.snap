// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue should render 1`] = `
<nav class="Menu fixed bottom-0 left-0 right-0 lg:top-0 lg:bottom-auto flex items-center justify-between flex-wrap bg-teal-500 p-6 z-50 Menu--hide">
  <div class="flex items-center flex-shrink-0 text-white mr-6"><span class="font-semibold text-xl tracking-tight">Condorcet</span></div>
  <div class="block lg:hidden"><button data-testid="burger-button" class="flex items-center px-3 py-2 border rounded text-teal-200 border-teal-400 hover:text-white hover:border-white"><svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" class="fill-current h-3 w-3">
        <title>Menu</title>
        <path d="M0 3h20v2H0V3zm0 6h20v2H0V9zm0 6h20v2H0v-2z"></path>
      </svg></button></div>
  <div class="justify-end w-full block lg:flex lg:items-center lg:w-auto">
    <div class="text-sm lg:flex-grow">
      <router-link-stub data-testid="navlink-user-polls" to="/polls" class="block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white">Mes scrutins</router-link-stub>
      <router-link-stub data-testid="navlink-public-polls" to="/polls/public" class="self-end block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white">Découvrir</router-link-stub>
      <router-link-stub data-testid="navlink-poll-creation" to="/polls/new" class="self-end block mt-4 mx-6 lg:inline-block lg:mt-0 text-teal-200 hover:text-white">Créer un scrutin</router-link-stub>
    </div>
    <div><a href="#" class="inline-block text-sm px-4 py-2 leading-none border rounded text-white border-white hover:border-transparent hover:text-teal-500 hover:bg-white mt-4 lg:mt-0">Deconnexion</a></div>
  </div>
</nav>
`;
